src/node_crypto.cc

```diff
@@ -3618,22 +3618,20 @@ void Sign::SignFinal(const FunctionCallbackInfo<Value>& args) {
   int salt_len = args[3].As<Int32>()->Value();
 
   ClearErrorOnReturn clear_error_on_return;
-  unsigned char md_value[8192];
-  unsigned int md_len = sizeof(md_value);
+  MallocedBuffer<unsigned char> sig;
```

Comment:
If the `MallocedBuffer` abstraction fails to fit, maybe create an `std::vector<unsigned char>` here and conditionally fill it in `SignFinal`.
If you know the maximal size you could even use an `std::array<unsigned char>` and save all the memory allocations.

Reply:
@refack That would undo the purpose of this change, which is to avoid extra allocations

Reply:
1. If we can do this on the stack it's free.
2. If we use a vector with move semantics, there's no allocation until we resize.
3. As you mentioned, this change benefit is readability.

And we have several guidelines that this breaks:
Google's:
1. [Local Variables](https://google.github.io/styleguide/cppguide.html#Local_Variables) - Place a function's variables in the narrowest scope possible, and initialize variables in the declaration.
2. [Output Parameters](https://google.github.io/styleguide/cppguide.html#Output_Parameters) - Prefer using return values rather than output parameters. If output-only parameters are used they should appear after input parameters.
3. [C++11](https://google.github.io/styleguide/cppguide.html#C++11) Use libraries and language extensions from C++11 when appropriate.

C++CG:
1. [ES.20](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-always): Always initialize an object
2. [F.20](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f20-for-out-output-values-prefer-return-values-to-output-parameters): For “out” output values, prefer return values to output parameters
3. [ES.1](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lib): Prefer the standard library to other libraries and to “handcrafted code”

Reply:
> Local Variables - Place a function's variables in the narrowest scope possible, and initialize variables in the declaration.

I don’t see that being broken here.

> Output Parameters - Prefer using return values rather than output parameters. If output-only parameters are used they should appear after input parameters.

That would go away if we implemented a `std::pair` return type, as you suggested. I :+1:’ed that comment.

> C++11 Use libraries and language extensions from C++11 when appropriate.

I don’t see that being broken here.

> ES.20: Always initialize an object

I don’t see that being broken here. Again, the rule explicitly lists it as a valid exception if the object is about to be initialized. 

> F.20: For “out” output values, prefer return values to output parameters

(same as above, `std::pair` would “solve” this)

Reply:
Yes, if we use a `pair` we get rid of most of the anti-patterns. That's my requested-change.

As for whether `MallocedBuffer` is the best abstraction for this case, that's an open question I pose to @tniessen.

IMHO it breaks since we need to resize in:
https://github.com/nodejs/node/blob/e5bd9d0f90fd95b631fb05555b3cd03261e7936c/src/node_crypto.cc#L3542
And a resizable contiguous container is literally the definition of `std::vector`. AFAIU since it's a part of the language compilers can minimize heap allocations